src/tracing/trace_flamegraph.c

```diff
@@ -0,0 +1,263 @@
+
+	if (tmp_flamegraph_context) {
+		tmp_flamegraph_context->mode = XDEBUG_TRACE_OPTION_FLAMEGRAPH_MEM;
+		tmp_flamegraph_context->stack = fg_stack_create();
+	}
+
+	return tmp_flamegraph_context;
+}
+
+void xdebug_trace_flamegraph_deinit(void *ctxt)
+{
+	xdebug_trace_flamegraph_context *context = (xdebug_trace_flamegraph_context*) ctxt;
+
+	if (NULL != context->stack) {
+		fg_stack_free(context->stack);
+	}
+
+	fclose(context->trace_file);
+	context->trace_file = NULL;
+	xdfree(context->trace_filename);
+
+	xdfree(context);
+}
+
+void xdebug_trace_flamegraph_write_header(void *ctxt)
+{
+}
+
+void xdebug_trace_flamegraph_write_footer(void *ctxt)
+{
+}
+
+char *xdebug_trace_flamegraph_get_filename(void *ctxt)
+{
+	xdebug_trace_flamegraph_context *context = (xdebug_trace_flamegraph_context*) ctxt;
+
+	return context->trace_filename;
+}
+
+void xdebug_trace_flamegraph_function_entry(void *ctxt, function_stack_entry *fse, int function_nr)
+{
+	xdebug_trace_flamegraph_context *context = (xdebug_trace_flamegraph_context*) ctxt;
+	flamegraph_stack_item *head = fg_stack_get_head(context->stack);
+	char *tmp_name, *prefix;
+
+	tmp_name = xdebug_show_fname(fse->function, 0, 0);
+
+	/* Insert an item into the custom stack which yields path prefix and
+	   a 0 value for the current function, if this function calls children
+	   they will append their costs to it. Once we reach out this function
+	   we will remove the element from the stack. */
+	if (NULL == head) {
+		/* Root function. */
+		fg_stack_push(context->stack, tmp_name);
+	} else {
+		prefix = xdmalloc(strlen(head->prefix) + 1 + strlen(tmp_name));
+		sprintf(prefix, "%s;%s", head->prefix, tmp_name);
+		fg_stack_push(context->stack, prefix);
+		xdfree(prefix);
+	}
+
+	xdfree(tmp_name);
+}
+
+void xdebug_trace_flamegraph_function_exit(void *ctxt, function_stack_entry *fse, int function_nr)
+{
+	xdebug_trace_flamegraph_context *context = (xdebug_trace_flamegraph_context*) ctxt;
+	xdebug_str str = XDEBUG_STR_INITIALIZER;
+	int inclusive, self;
+	flamegraph_stack_item *stack_item = fg_stack_pop(context->stack);
+
+	if (NULL == stack_item) {
+		/* This should never happen, better be safe than sorry. */
+		/* @todo log error? */
+		return;
+	}
+
+	inclusive = compute_inclusive_value(context, fse);
+	self = inclusive - stack_item->value;
+	xdebug_str_add_fmt(&str, "%s %d\n", stack_item->prefix, self);
+	xdfree(stack_item->prefix);
+	xdfree(stack_item);
+
+	/* Increment head value (which is now parent) by inclusive cost. */
+	fg_stack_inc(context->stack, inclusive);
+
+	fprintf(context->trace_file, "%s", str.d);
+	fflush(context->trace_file);
+	xdfree(str.d);
+}
+
+void xdebug_trace_flamegraph_function_return_value(void *ctxt, function_stack_entry *fse, int function_nr, zval *return_value)
+{
+}
+
+void xdebug_trace_flamegraph_assignment(void *ctxt, function_stack_entry *fse, char *full_varname, zval *retval, char *right_full_varname, const char *op, char *filename, int lineno)
+{
+}
```

Comment:
As as I said upper, I stumbled upon more than one handler that wasn't checked for NULL and created segfault. Since the topic of this PR is the flamegraph trace and I was unsure about your design choices, I kept those empty functions.

Reply:
Closing that for https://github.com/xdebug/xdebug/pull/761#discussion_r677636947